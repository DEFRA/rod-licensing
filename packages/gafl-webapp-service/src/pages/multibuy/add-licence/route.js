import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { ADD_LICENCE, CONTACT_SUMMARY, LICENCE_SUMMARY } from '../../../uri.js'
import { nextPage } from '../../../routes/next-page.js'

import GetDataRedirect from '../../../handlers/get-data-redirect.js'

export const getData = async request => {
  const status = await request.cache().helpers.status.getCurrentPermission()

  if (!status[LICENCE_SUMMARY.page]) {
    throw new GetDataRedirect(LICENCE_SUMMARY.uri)
  }

  if (!status[CONTACT_SUMMARY.page]) {
    throw new GetDataRedirect(CONTACT_SUMMARY.uri)
  }

  const transaction = await request.cache().helpers.transaction.get()

  return { numberOfLicences: transaction.permissions.length }
}

const validator = Joi.object({
  'add-licence': Joi.string()
    .valid('yes', 'no')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

export default pageRoute(ADD_LICENCE.page, ADD_LICENCE.uri, validator, nextPage, getData)
