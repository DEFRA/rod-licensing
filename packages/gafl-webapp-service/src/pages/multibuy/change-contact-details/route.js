import pageRoute from '../../../routes/page-route.js'
import GetDataRedirect from '../../../handlers/get-data-redirect.js'
import { countries } from '../../../processors/refdata-helper.js'
import { HOW_CONTACTED } from '../../../processors/mapping-constants.js'
import { CHANGE_CONTACT_DETAILS_SEEN } from '../../../constants.js'
import { isPhysical } from '../../../processors/licence-type-display.js'
import { nextPage } from '../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../processors/uri-helper.js'

import {
  CHANGE_CONTACT_DETAILS,
  ADDRESS_ENTRY,
  ADDRESS_SELECT,
  ADDRESS_LOOKUP,
  CONTACT,
  NEWSLETTER,
  LICENCE_FULFILMENT,
  LICENCE_CONFIRMATION_METHOD,
  CHANGE_LICENCE_OPTIONS
} from '../../../uri.js'

const CONTACT_TEXT_DEFAULT = {
  EMAIL: 'contact_summary_email',
  TEXT: 'contact_summary_text_sngl',
  DEFAULT: 'contact_summary_license_default'
}

const CONTACT_TEXT_NON_PHYSICAL = {
  EMAIL: 'contact_summary_email',
  TEXT: 'contact_summary_text_plrl',
  DEFAULT: 'contact_summary_license_non_physical'
}

const CONTACT_TEXT_PHYSICAL = {
  EMAIL: 'contact_summary_email',
  TEXT: 'contact_summary_text_plrl',
  DEFAULT: 'contact_summary_license_physical'
}

const CHANGE_CONTACT = 'change-contact'

class RowGenerator {
  constructor (request, permission) {
    this.request = request
    this.permission = permission
    this.labels = request.i18n.getCatalog()
  }

  _getPreferredMethodOfReminderText (contactTextSpec) {
    switch (this.permission.licensee.preferredMethodOfReminder) {
      case HOW_CONTACTED.email:
        return `${this.labels[contactTextSpec.EMAIL]}${this.permission.licensee.email}`
      case HOW_CONTACTED.text:
        return `${this.labels[contactTextSpec.TEXT]}${this.permission.licensee.mobilePhone}`
      default:
        return this.labels[contactTextSpec.DEFAULT]
    }
  }

  _getPreferredMethodOfConfirmation (contactTextSpec) {
    switch (this.permission.licensee.preferredMethodOfConfirmation) {
      case HOW_CONTACTED.email:
        return `${this.labels[contactTextSpec.EMAIL]}${this.permission.licensee.email}`
      case HOW_CONTACTED.text:
        return `${this.labels[contactTextSpec.TEXT]}${this.permission.licensee.mobilePhone}`
      default:
        return this.labels[contactTextSpec.DEFAULT]
    }
  }

  _generateRow (label, text, rawHref, visuallyHiddenText, id) {
    const href = addLanguageCodeToUri(this.request, rawHref)
    return {
      key: {
        text: label
      },
      value: {
        text
      },
      ...(href && {
        actions: {
          items: [
            {
              href,
              visuallyHiddenText,
              text: this.labels.contact_summary_change,
              attributes: { id }
            }
          ]
        }
      })
    }
  }

  generateStandardRow (label, text, rawHref, visuallyHiddenText, id) {
    return this._generateRow(this.labels[label], this.labels[text], rawHref, this.labels[visuallyHiddenText], id)
  }

  generateAddressRow (countryName) {
    const { licensee } = this.permission
    const text = [licensee.premises, licensee.street, licensee.locality, licensee.town, licensee.postcode, countryName?.toUpperCase()]
      .filter(Boolean)
      .join(', ')

    return this._generateRow(
      this.labels.contact_summary_row_address,
      text,
      ADDRESS_LOOKUP.uri,
      this.labels.contact_summary_hidden_address,
      'change-address'
    )
  }

  generateContactRow (label, href, visuallyHiddenText, id, contactTextSpec = CONTACT_TEXT_DEFAULT) {
    const contactText =
      label === 'contact_summary_row_contact'
        ? this._getPreferredMethodOfReminderText(contactTextSpec)
        : this._getPreferredMethodOfConfirmation(contactTextSpec)
    return this._generateRow(this.labels[label], contactText, href, this.labels[visuallyHiddenText], id)
  }
}

const checkNavigation = (status, permission) => {
  if (!permission.isRenewal) {
    if (!status[ADDRESS_ENTRY.page] && !status[ADDRESS_SELECT.page]) {
      throw new GetDataRedirect(ADDRESS_LOOKUP.uri)
    }

    if (!status[CONTACT.page]) {
      throw new GetDataRedirect(CONTACT.uri)
    }
  }

  if (isPhysical(permission)) {
    if (!status[LICENCE_FULFILMENT.page]) {
      throw new GetDataRedirect(LICENCE_FULFILMENT.uri)
    }
    if (!status[LICENCE_CONFIRMATION_METHOD.page]) {
      throw new GetDataRedirect(LICENCE_CONFIRMATION_METHOD.uri)
    }
  }
}

const getLicenseeDetailsSummaryRows = (permission, countryName, request) => {
  const rowGenerator = new RowGenerator(request, permission)

  const licenseeSummaryArray = [rowGenerator.generateAddressRow(countryName)]
  if (isPhysical(permission)) {
    if (permission.licensee.postalFulfilment) {
      licenseeSummaryArray.push(
        rowGenerator.generateStandardRow(
          'contact_summary_row_licence',
          'contact_summary_license_physical',
          LICENCE_FULFILMENT.uri,
          'contact_summary_hidden_licence_fulfilment',
          'change-licence-fulfilment-option'
        ),
        rowGenerator.generateContactRow(
          'contact_summary_row_licence_conf',
          LICENCE_CONFIRMATION_METHOD.uri,
          'contact_summary_hidden_licence_confirmation',
          'change-licence-confirmation-option'
        )
      )
    } else {
      licenseeSummaryArray.push(
        rowGenerator.generateContactRow(
          'contact_summary_row_licence',
          LICENCE_FULFILMENT.uri,
          'contact_summary_hidden_licence_confirmation',
          'change-licence-confirmation-option'
        )
      )
    }

    licenseeSummaryArray.push(
      rowGenerator.generateContactRow(
        'contact_summary_row_contact',
        CONTACT.uri,
        'contact_summary_hidden_contact',
        CHANGE_CONTACT,
        CONTACT_TEXT_PHYSICAL
      )
    )
  } else {
    licenseeSummaryArray.push(
      rowGenerator.generateContactRow(
        'contact_summary_row_licence_details',
        CONTACT.uri,
        'contact_summary_hidden_contact',
        CHANGE_CONTACT,
        CONTACT_TEXT_NON_PHYSICAL
      )
    )
  }

  if (permission.isLicenceForYou) {
    const text = permission.licensee.preferredMethodOfNewsletter === HOW_CONTACTED.none ? 'no' : 'yes'
    licenseeSummaryArray.push(
      rowGenerator.generateStandardRow(
        'contact_summary_row_newsletter',
        text,
        NEWSLETTER.uri,
        'contact_summary_hidden_newsletter',
        'change-newsletter'
      )
    )
  }

  return licenseeSummaryArray
}

const getData = async request => {
  const status = await request.cache().helpers.status.getCurrentPermission()
  const permission = await request.cache().helpers.transaction.getCurrentPermission()
  const mssgs = request.i18n.getCatalog()

  checkNavigation(status, permission)

  status.changeContactDetails = CHANGE_CONTACT_DETAILS_SEEN.SEEN

  console.log(status.changeContactDetails)

  await request.cache().helpers.status.setCurrentPermission(status)
  const countryName = await countries.nameFromCode(permission.licensee.countryCode)

  const changeLicenceDetails = permission.isLicenceForYou ? mssgs.change_licence_details_you : mssgs.change_licence_details_other

  return {
    summaryTable: getLicenseeDetailsSummaryRows(permission, countryName, request),
    uri: {
      changeLicenceOptions: CHANGE_LICENCE_OPTIONS.uri
    },
    changeLicenceDetails
  }
}

export default pageRoute(CHANGE_CONTACT_DETAILS.page, CHANGE_CONTACT_DETAILS.uri, null, nextPage, getData)
