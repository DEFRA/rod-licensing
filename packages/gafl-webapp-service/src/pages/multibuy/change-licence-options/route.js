import pageRoute from '../../../routes/page-route.js'
import GetDataRedirect from '../../../handlers/get-data-redirect.js'
import { licenceTypeDisplay, getErrorPage } from '../../../processors/licence-type-display.js'
import { displayStartTime, cacheDateFormat } from '../../../processors/date-and-time-display.js'
import findPermit from '../../summary/find-permit.js'
import { CHANGE_LICENCE_OPTIONS_SEEN } from '../../../constants.js'
import moment from 'moment-timezone'
import {
  NAME,
  CHANGE_LICENCE_OPTIONS,
  CHANGE_CONTACT_DETAILS,
  LICENCE_LENGTH,
  LICENCE_TYPE,
  LICENCE_TO_START,
  DISABILITY_CONCESSION,
  DATE_OF_BIRTH,
  RENEWAL_START_DATE,
  NEW_TRANSACTION
} from '../../../uri.js'
import { nextPage } from '../../../routes/next-page.js'
import { START_AFTER_PAYMENT_MINUTES, SERVICE_LOCAL_TIME } from '@defra-fish/business-rules-lib'
import { CONCESSION, CONCESSION_PROOF } from '../../../processors/mapping-constants.js'
import * as concessionHelper from '../../../processors/concession-helper.js'

export const getData = async request => {
  const status = await request.cache().helpers.status.getCurrentPermission()
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  if (!status.renewal) {
    /*
     * Before we try and filter the permit it is necessary to check that the user has navigated through
     * the journey in such a way as to have gather all the required data. They may have manipulated the
     * journey by typing into the address bar in which case they will be redirected back to the
     * appropriate point in the journey. For a renewal this is not necessary.
     */
    const errorOnPage = getErrorPage(permission)
    if (errorOnPage) {
      throw new GetDataRedirect(errorOnPage)
    }
  }

  status.fromLicenceOptions = CHANGE_LICENCE_OPTIONS_SEEN.SEEN
  await request.cache().helpers.status.setCurrentPermission(status)
  await findPermit(permission, request)
  const startTimeString = displayStartTime(request, permission)

  return {
    permission,
    startTimeString,
    startAfterPaymentMinutes: START_AFTER_PAYMENT_MINUTES,
    licenceTypeStr: licenceTypeDisplay(permission, request.i18n.getCatalog()),
    isRenewal: status.renewal,
    isContinuing: !!(permission.renewedEndDate && permission.renewedEndDate === permission.licenceStartDate),
    hasExpired: moment(moment().tz(SERVICE_LOCAL_TIME)).isAfter(moment(permission.renewedEndDate, cacheDateFormat)),
    disabled: permission.concessions && permission.concessions.find(c => c.type === CONCESSION.DISABLED),
    concessionProofs: CONCESSION_PROOF,
    hasJunior: concessionHelper.hasJunior(permission),
    cost: permission.permit.cost,
    birthDateStr: moment(permission.licensee.birthDate, cacheDateFormat).format('Do MMMM YYYY'),
    uri: {
      name: NAME.uri,
      licenceLength: LICENCE_LENGTH.uri,
      licenceType: LICENCE_TYPE.uri,
      licenceToStart: LICENCE_TO_START.uri,
      dateOfBirth: DATE_OF_BIRTH.uri,
      disabilityConcession: DISABILITY_CONCESSION.uri,
      licenceStartDate: status.renewal ? RENEWAL_START_DATE.uri : LICENCE_TO_START.uri,
      clear: NEW_TRANSACTION.uri,
      changeContactDetails: CHANGE_CONTACT_DETAILS.uri
    }
  }
}

export default pageRoute(CHANGE_LICENCE_OPTIONS.page, CHANGE_LICENCE_OPTIONS.uri, null, nextPage, getData)
