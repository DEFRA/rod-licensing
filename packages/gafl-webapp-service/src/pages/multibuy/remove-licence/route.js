import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { REMOVE_LICENCE, VIEW_LICENCES } from '../../../uri.js'
import { nextPage } from '../../../routes/next-page.js'
import { licenceTypeDisplay, licenceTypeAndLengthDisplay } from '../../../processors/licence-type-display.js'
import { displayStartTime } from '../../../processors/date-and-time-display.js'
import { addLanguageCodeToUri } from '../../../processors/uri-helper.js'

export const getData = async request => {
  await request.cache().helpers.status.set({ currentPermissionIdx: parseInt(request.query.permissionIndex) })
  const permission = await request.cache().helpers.transaction.getCurrentPermission()
  const mssgs = request.i18n.getCatalog()

  return {
    licenceHolder: `${permission.licensee.firstName} ${permission.licensee.lastName}`,
    type: licenceTypeDisplay(permission, mssgs),
    length: licenceTypeAndLengthDisplay(permission, mssgs),
    start: displayStartTime(request, permission),
    price: permission.permit.cost,
    uri: {
      view_licences: addLanguageCodeToUri(request, VIEW_LICENCES.uri)
    }
  }
}

export const validator = Joi.object({}).options({ abortEarly: false, allowUnknown: true })

export default pageRoute(REMOVE_LICENCE.page, REMOVE_LICENCE.uri, validator, nextPage, getData)
