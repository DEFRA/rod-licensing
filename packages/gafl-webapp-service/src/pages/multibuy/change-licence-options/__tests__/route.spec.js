import { getData, checkNavigation } from '../route'
import {
  DATE_OF_BIRTH,
  LICENCE_LENGTH,
  LICENCE_TO_START,
  LICENCE_TYPE as LICENCE_TYPE_URI,
  NAME,
  RENEWAL_START_DATE
} from '../../../../uri.js'
import GetDataRedirect from '../../../../handlers/get-data-redirect.js'
import { CONCESSION } from '../../../../processors/mapping-constants.js'

jest.mock('../../../summary/find-permit.js')
jest.mock('../../../../processors/mapping-constants.js', () => ({
  CONCESSION: {
    SENIOR: 'Senior person',
    JUNIOR: 'Junior person',
    DISABLED: 'Disabled person'
  },
  LICENCE_TYPE: {
    'trout-and-coarse': 'Trout and coarse',
    'salmon-and-sea-trout': 'Salmon and sea trout'
  },
  CONCESSION_PROOF: {
    NI: 'NIN',
    blueBadge: 'BB',
    none: 'Not Proof'
  }
}))

describe('change-licence-options > route', () => {
  beforeEach(jest.clearAllMocks)

  describe('getData', () => {
    const mockStatusCacheGet = jest.fn(() => ({}))
    const mockStatusCacheSet = jest.fn()
    const mockTransactionCacheGet = jest.fn()
    const mockTransactionCacheSet = jest.fn()

    const mockRequest = () => ({
      cache: () => ({
        helpers: {
          status: {
            getCurrentPermission: mockStatusCacheGet,
            setCurrentPermission: mockStatusCacheSet
          },
          transaction: {
            get: jest.fn(),
            set: jest.fn(),
            getCurrentPermission: mockTransactionCacheGet,
            setCurrentPermission: mockTransactionCacheSet
          }
        }
      })
    })

    it('test output of getData', async () => {
      mockTransactionCacheGet.mockImplementationOnce(() => ({
        birthDateStr: '1st January 1946',
        concessionProofs: {
          NI: 'National Insurance Number',
          blueBadge: 'Blue Badge',
          none: 'No Proof'
        },
        cost: 6,
        disabled: true,
        hasExpired: false,
        hasJunior: false,
        isContinuing: false,
        isRenewal: true,
        licenceLength: '12M',
        licenceStartDate: '2021-07-01',
        licenceType: 'Salmon and sea trout',
        licenceTypeStr: 'Salmon and sea trout',
        licensee: {
          birthDate: '1946-01-01',
          firstName: 'Graham',
          lastName: 'Willis'
        },
        numberOfRods: '3',
        permit: {
          cost: 6
        },
        startAfterPaymentMinute: 30,
        startTimeString: '0.00am (first minute of the day) on 1 July 2021',
        uri: {
          clear: '/buy/new',
          dateOfBirth: '/buy/date-of-birth',
          disabilityConcession: '/buy/disability-concession',
          licenceLength: '/buy/licence-length',
          licenceStartDate: '/buy/start-kind',
          licenceToStart: '/buy/start-kind',
          licenceType: '/buy/licence-type',
          name: '/buy/name'
        }
      }))
      const result = await getData(mockRequest())
      expect(result).toMatchSnapshot()
    })

    it('should return a redirect error if firstName is not included on the licensee', async () => {
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({ licensee: {} }))
      const getDataRedirectError = new GetDataRedirect(NAME.uri)
      await expect(getData(mockRequest(mockTransaction))).rejects.toThrow(getDataRedirectError)
    })

    it('should return a redirect error if lastName is not included on the licensee', async () => {
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({ licensee: { firstName: 'John' } }))
      const getDataRedirectError = new GetDataRedirect(NAME.uri)
      await expect(getData(mockRequest(mockTransaction))).rejects.toThrow(getDataRedirectError)
    })

    it('should return a redirect error if date of birth is not included on the licensee', async () => {
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({ licensee: { firstName: 'John', lastName: 'Smith' } }))
      const getDataRedirectError = new GetDataRedirect(DATE_OF_BIRTH.uri)
      await expect(getData(mockRequest(mockTransaction))).rejects.toThrow(getDataRedirectError)
    })

    it('should return a redirect error if start date is not included on the licensee', async () => {
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        licenceStartDate: null
      }))
      const getDataRedirectError = new GetDataRedirect(LICENCE_TO_START.uri)
      await expect(getData(mockRequest(mockTransaction))).rejects.toThrow(getDataRedirectError)
    })

    it('should return a redirect error if number of rods is not included on the licensee', async () => {
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        licenceStartDate: '2025-01-01',
        licenceType: 'Salmon and sea trout'
      }))
      const getDataRedirectError = new GetDataRedirect(LICENCE_TYPE_URI.uri)
      await expect(getData(mockRequest(mockTransaction))).rejects.toThrow(getDataRedirectError)
    })

    it('should return a redirect error if licence type is not included on the licensee', async () => {
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        licenceStartDate: '2025-01-01',
        numberOfRods: '3'
      }))
      const getDataRedirectError = new GetDataRedirect(LICENCE_TYPE_URI.uri)
      await expect(getData(mockRequest(mockTransaction))).rejects.toThrow(getDataRedirectError)
    })

    it('should return a redirect error if licence length is not included on the licensee', async () => {
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        licenceStartDate: '2025-01-01',
        numberOfRods: '3',
        licenceType: 'Salmon and sea trout'
      }))
      const getDataRedirectError = new GetDataRedirect(LICENCE_LENGTH.uri)
      await expect(getData(mockRequest(mockTransaction))).rejects.toThrow(getDataRedirectError)
    })

    it.each([
      ['Salmon and sea trout', '3', 'Salmon and sea trout'],
      ['Trout and coarse', '2', 'Trout and coarse, up to 2 rods'],
      ['Trout and coarse', '3', 'Trout and coarse, up to 3 rods']
    ])('should return licenceTypeStr as the same value as licenceTypeDisplay', async (type, numOfRods, expectedResult) => {
      const mockStatus = mockStatusCacheGet.mockImplementationOnce(() => ({ renewal: false }))
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licenceStartDate: mockStatus.renewal ? RENEWAL_START_DATE.uri : LICENCE_TO_START.uri,
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        licenceLength: '12M',
        permit: {
          cost: 6
        },
        licenceType: type,
        numberOfRods: numOfRods
      }))
      const result = await getData(mockRequest(mockTransaction))
      expect(result.licenceTypeStr).toBe(expectedResult)
    })

    it.each([
      ['Junior', 'Salmon and sea trout', '3', 'Salmon and sea trout'],
      ['Junior', 'Trout and coarse', '2', 'Trout and coarse, up to 2 rods'],
      ['Junior', 'Trout and coarse', '3', 'Trout and coarse, up to 3 rods'],
      ['Senior', 'Salmon and sea trout', '3', 'Salmon and sea trout'],
      ['Senior', 'Trout and coarse', '2', 'Trout and coarse, up to 2 rods'],
      ['Senior', 'Trout and coarse', '3', 'Trout and coarse, up to 3 rods']
    ])(
      'should return same result as concessionHelper, if licenceType includes concession hasJunior or hasSenior',
      async (concession, type, numOfRods, expectedResult) => {
        const request = {
          cache: () => ({
            helpers: {
              status: { getCurrentPermission: () => mockStatus },
              transaction: { getCurrentPermission: () => mockTransaction }
            }
          })
        }
        const mockStatus = mockStatusCacheGet.mockImplementationOnce(() => ({ renewal: false }))
        const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
          licenceStartDate: mockStatus.renewal ? RENEWAL_START_DATE.uri : LICENCE_TO_START.uri,
          licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
          licenceLength: '12M',
          permit: {
            cost: 6
          },
          concessions: {
            concession,
            find: jest.fn(c => {
              return c.type === 'disabled'
            })
          },
          licenceType: type,
          numberOfRods: numOfRods
        }))
        const result = await getData(mockRequest(request))
        expect(result.licenceTypeStr).toBe(expectedResult)
      }
    )

    it('should return isContinuing as true if renewedEndDate is equal to licenceStartDate', async () => {
      const request = {
        cache: () => ({
          helpers: {
            status: { getCurrentPermission: () => mockStatus },
            transaction: { getCurrentPermission: () => mockTransaction }
          }
        })
      }
      const mockStatus = mockStatusCacheGet.mockImplementationOnce(() => ({ renewal: true }))
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        permit: {
          cost: 6
        },
        renewedEndDate: '2022-06-22T14:48:00.000Z',
        licenceStartDate: '2022-06-22T14:48:00.000Z'
      }))
      const result = await getData(mockRequest(request))
      expect(result.isContinuing).toBeTruthy()
    })

    it('should return isContinuing as false if both renewedEndDate is not equal to licenceStartDate', async () => {
      const request = {
        cache: () => ({
          helpers: {
            status: { getCurrentPermission: () => mockStatus },
            transaction: { getCurrentPermission: () => mockTransaction }
          }
        })
      }
      const mockStatus = mockStatusCacheGet.mockImplementationOnce(() => ({ renewal: true }))
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        permit: {
          cost: 6
        },
        renewedEndDate: '2022-06-22T14:48:00.000Z',
        licenceStartDate: '2022-07-22T15:48:00.000Z'
      }))
      const result = await getData(mockRequest(request))
      expect(result.isContinuing).toBeFalsy()
    })

    it('should return true if a concession has been selected and only if disabled', async () => {
      const request = {
        cache: () => ({
          helpers: {
            status: { getCurrentPermission: () => mockStatus },
            transaction: { getCurrentPermission: () => mockTransaction }
          }
        })
      }
      const mockStatus = mockStatusCacheGet.mockImplementationOnce(() => ({ renewal: true }))
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        permit: {
          cost: 6
        },
        concessions: [
          {
            type: CONCESSION.DISABLED
          }
        ]
      }))
      const result = await getData(mockRequest(request))
      expect(result.disabled).toBeTruthy()
    })

    it('should return RENEWAL_START_DATE uri if renewal is true', async () => {
      const request = {
        cache: () => ({
          helpers: {
            status: { getCurrentPermission: () => mockStatus },
            transaction: { getCurrentPermission: () => mockTransaction }
          }
        })
      }
      const mockStatus = mockStatusCacheGet.mockImplementationOnce(() => ({ renewal: true }))
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licenceStartDate: mockStatus.renewal ? RENEWAL_START_DATE.uri : LICENCE_TO_START.uri,
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        numberOfRods: '3',
        licenceLength: '12M',
        permit: {
          cost: 6
        },
        licenceType: 'Salmon and sea trout'
      }))
      const result = await getData(mockRequest(request))
      expect(result.uri.licenceStartDate).toBe(RENEWAL_START_DATE.uri)
    })

    it('should return LICENCE_TO_START uri if renewal is false', async () => {
      const request = {
        cache: () => ({
          helpers: {
            status: { getCurrentPermission: () => mockStatus },
            transaction: { getCurrentPermission: () => mockTransaction }
          }
        })
      }
      const mockStatus = mockStatusCacheGet.mockImplementationOnce(() => ({ renewal: false }))
      const mockTransaction = mockTransactionCacheGet.mockImplementationOnce(() => ({
        licenceStartDate: mockStatus.renewal ? RENEWAL_START_DATE.uri : LICENCE_TO_START.uri,
        licensee: { firstName: 'John', lastName: 'Smith', birthDate: '1996-01-01' },
        numberOfRods: '3',
        licenceLength: '12M',
        permit: {
          cost: 6
        },
        licenceType: 'Salmon and sea trout'
      }))
      const result = await getData(mockRequest(request))
      expect(result.uri.licenceStartDate).toBe(LICENCE_TO_START.uri)
    })
  })

  describe('CheckNavigation', () => {
    it('should return a redirect error if firstName is not included on the licensee', async () => {
      const permission = { licensee: { firstName: null } }
      expect(() => checkNavigation(permission)).toThrow(GetDataRedirect)
    })

    it('should return a redirect error if lastName is not included on the licensee', async () => {
      const permission = { licensee: { firstName: 'Scott', lastName: null } }
      expect(() => checkNavigation(permission)).toThrow(GetDataRedirect)
    })

    it('should return a redirect error if date of birth is not included on the licensee', async () => {
      const permission = { licensee: { firstName: 'Scott', lastName: 'Michael', birthDate: null } }
      expect(() => checkNavigation(permission)).toThrow(GetDataRedirect)
    })

    it('should return a redirect error if start date is not included on the licensee', async () => {
      const permission = {
        licensee: { firstName: 'Scott', lastName: 'Michael', birthDate: '1996-01-01' },
        licenceStartDate: null
      }
      expect(() => checkNavigation(permission)).toThrow(GetDataRedirect)
    })

    it('should return a redirect error if number of rods is not included on the licensee', async () => {
      const permission = {
        licensee: { firstName: 'Scott', lastName: 'Michael', birthDate: '1996-01-01' },
        licenceStartDate: '2022-07-01',
        numberOfRods: null
      }
      expect(() => checkNavigation(permission)).toThrow(GetDataRedirect)
    })

    it('should return a redirect error if licence type is not included on the licensee', async () => {
      const permission = {
        licensee: { firstName: 'Scott', lastName: 'Michael', birthDate: '1996-01-01' },
        licenceStartDate: '2022-07-01',
        numberOfRods: 3,
        licenceType: null
      }
      expect(() => checkNavigation(permission)).toThrow(GetDataRedirect)
    })

    it('should return a redirect error if licence length is not included on the licensee', async () => {
      const permission = {
        licensee: { firstName: 'Scott', lastName: 'Michael', birthDate: '1996-01-01' },
        licenceStartDate: '2022-07-01',
        numberOfRods: 3,
        licenceType: 'Salmon and sea trout',
        licenceLength: undefined
      }
      expect(() => checkNavigation(permission)).toThrow(GetDataRedirect)
    })
  })
})
