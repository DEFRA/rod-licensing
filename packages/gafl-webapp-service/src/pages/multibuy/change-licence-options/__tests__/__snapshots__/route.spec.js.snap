// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change-licence-options > route getData test output of getData for a continuing, expired permission 1`] = `
Object {
  "birthDateStr": "1st January 1946",
  "concessionProofs": Object {
    "NI": "NIN",
    "blueBadge": "BB",
    "none": "Not Proof",
  },
  "cost": 6,
  "disabled": undefined,
  "hasExpired": true,
  "hasJunior": undefined,
  "isContinuing": true,
  "isRenewal": undefined,
  "licenceTypeStr": Symbol(Salmon and sea trout),
  "permission": Object {
    "licenceStartDate": "2021-07-01",
    "licenceType": "Salmon and sea trout",
    "licensee": Object {
      "birthDate": "1946-01-01",
    },
    "numberOfRods": "3",
    "permit": Object {
      "cost": 6,
    },
    "renewedEndDate": "2021-07-01",
  },
  "startAfterPaymentMinutes": 30,
  "startTimeString": "Test on 1 July 2021",
  "uri": Object {
    "clear": "/buy/new",
    "dateOfBirth": "/buy/date-of-birth",
    "disabilityConcession": "/buy/disability-concession",
    "licenceLength": "/buy/licence-length",
    "licenceStartDate": "/buy/start-kind",
    "licenceToStart": "/buy/start-kind",
    "licenceType": "/buy/licence-type",
    "name": "/buy/name",
  },
}
`;

exports[`change-licence-options > route getData test output of getData for a disabled permission 1`] = `
Object {
  "birthDateStr": "1st January 1946",
  "concessionProofs": Object {
    "NI": "NIN",
    "blueBadge": "BB",
    "none": "Not Proof",
  },
  "cost": 6,
  "disabled": Object {
    "type": "Disabled person",
  },
  "hasExpired": true,
  "hasJunior": false,
  "isContinuing": false,
  "isRenewal": undefined,
  "licenceTypeStr": Symbol(Salmon and sea trout),
  "permission": Object {
    "concessions": Array [
      Object {
        "type": "Disabled person",
      },
    ],
    "licenceStartDate": "2021-07-01",
    "licenceType": "Salmon and sea trout",
    "licensee": Object {
      "birthDate": "1946-01-01",
    },
    "numberOfRods": "3",
    "permit": Object {
      "cost": 6,
    },
    "renewedEndDate": "2021-07-02",
  },
  "startAfterPaymentMinutes": 30,
  "startTimeString": "Test on 1 July 2021",
  "uri": Object {
    "clear": "/buy/new",
    "dateOfBirth": "/buy/date-of-birth",
    "disabilityConcession": "/buy/disability-concession",
    "licenceLength": "/buy/licence-length",
    "licenceStartDate": "/buy/start-kind",
    "licenceToStart": "/buy/start-kind",
    "licenceType": "/buy/licence-type",
    "name": "/buy/name",
  },
}
`;

exports[`change-licence-options > route getData test output of getData for a standard permission 1`] = `
Object {
  "birthDateStr": "1st January 1946",
  "concessionProofs": Object {
    "NI": "NIN",
    "blueBadge": "BB",
    "none": "Not Proof",
  },
  "cost": 6,
  "disabled": undefined,
  "hasExpired": true,
  "hasJunior": undefined,
  "isContinuing": false,
  "isRenewal": undefined,
  "licenceTypeStr": Symbol(Salmon and sea trout),
  "permission": Object {
    "licenceStartDate": "2021-07-01",
    "licenceType": "Salmon and sea trout",
    "licensee": Object {
      "birthDate": "1946-01-01",
    },
    "numberOfRods": "3",
    "permit": Object {
      "cost": 6,
    },
    "renewedEndDate": "2021-07-02",
  },
  "startAfterPaymentMinutes": 30,
  "startTimeString": "Test on 1 July 2021",
  "uri": Object {
    "clear": "/buy/new",
    "dateOfBirth": "/buy/date-of-birth",
    "disabilityConcession": "/buy/disability-concession",
    "licenceLength": "/buy/licence-length",
    "licenceStartDate": "/buy/start-kind",
    "licenceToStart": "/buy/start-kind",
    "licenceType": "/buy/licence-type",
    "name": "/buy/name",
  },
}
`;
